src/software/geom/spline.cpp

```diff
@@ -0,0 +1,79 @@
+#include "software/geom/spline.h"
+
+Spline::Spline(const std::vector<Point>& points)
+    : knots(points), domain(std::make_pair(0.0, 1.0))
+{
+    initLinearSegments(points);
+}
+
+Spline::Spline(const std::initializer_list<Point>& points)
+    : knots(points), domain(std::make_pair(0.0, 1.0))
+{
+    initLinearSegments(points);
+}
+
+Point Spline::valueAt(double val) const
+{
+    if (val < domain.first || val > domain.second)
+    {
+        throw std::invalid_argument("val is outside of the domain of this spline");
+    }
+
+    if (val == domain.first)
+    {
+        return knots.front();
+    }
+
+    if (val == domain.second)
+    {
+        return knots.back();
+    }
+
+    // assume that the segments are equally spaced in domain
```

Comment:
Why are we making this assumption?

Reply:
I haven't come up with an efficient/elegant way to check which segment a val corresponds to and it's not necessary for what we use splines for at the moment. Please suggest a nice way to check intervals other than iterating through the segments or if you think that's the way to go.

Reply:
Hmm, you could do a binary search? Check if the segment in the middle contains the value, and if it doesn't, check if it's greater or less then the values contained by the segment in the middle. If the value is greater search the top half of the list, otherwise search the bottom half of the list. Annnndddd _recurse_

Reply:
yeah I guess...